matchers/collections.go

```diff
@@ -0,0 +1,199 @@
+			// matched an item, we don't need to describe that matcher or that item.
+			elements, err := getSliceOrArrayElementValues(value)
+			if err != nil {
+				return err.Error()
+			}
+			type unmatchedElement struct {
+				index int
+				value interface{}
+			}
+			unmatchedElements := make([]unmatchedElement, 0)
+			for i, e := range elements {
+				unmatchedElements = append(unmatchedElements, unmatchedElement{index: i, value: e})
+			}
+			unmatchedMatchers := append([]Matcher(nil), matchers...)
+			for i := 0; i < len(unmatchedMatchers); i++ {
+				m := unmatchedMatchers[i]
+				for j := 0; j < len(unmatchedElements); j++ {
+					if m.test(unmatchedElements[j].value) {
+						unmatchedElements = append(unmatchedElements[0:j], unmatchedElements[j+1:]...)
+						unmatchedMatchers = append(unmatchedMatchers[0:i], unmatchedMatchers[i+1:]...)
+						j--
+						i--
+						break
+					}
+				}
+			}
+			if len(unmatchedMatchers) == 0 {
+				// Every matcher matched a value but there were some values left over.
+				parts := make([]string, 0)
+				for _, e := range unmatchedElements {
+					parts = append(parts, fmt.Sprintf("[%d]: %s", e.index, DescribeValue(e.value)))
+				}
+				return fmt.Sprintf("got more items than expected: %s", strings.Join(parts, ", "))
+			}
+			if len(unmatchedMatchers) == 1 && len(unmatchedElements) == 1 {
+				// In this case we'll assume that this was the element that was supposed to be matched
+				// by this matcher, but its value was wrong somehow. So we'll print the failure message
+				// from the matcher.
+				return fmt.Sprintf("failed expectation for one item [%d] with value: %s\nfailure was: %s",
+					unmatchedElements[0].index, DescribeValue(unmatchedElements[0].value),
+					unmatchedMatchers[0].describeFailure(unmatchedElements[0].value))
+			}
+			// If there was more than one unmatched matcher and/or element, we can't really guess
+			// which one was supposed to go with which, so we'll just print all the unmet conditions.
+			return fmt.Sprintf("no items were found to match: %s", describeMatchers(unmatchedMatchers, ", "))
+		},
+	)
+}
+
+func getSliceOrArrayElementValues(sliceValue interface{}) ([]interface{}, error) {
+	v := reflect.ValueOf(sliceValue)
+	if v.Type().Kind() != reflect.Slice && v.Type().Kind() != reflect.Array {
+		return nil, fmt.Errorf("expected slice or array value but got %T", sliceValue)
+	}
+	ret := make([]interface{}, 0, v.Len())
+	for i := 0; i < v.Len(); i++ {
+		ret = append(ret, v.Index(i).Interface())
+	}
+	return ret, nil
+}
+
+// ValueForKey is a MatcherTransform that takes a map, looks up a value in it by key,
+// and applies a matcher to that value. It fails if no such key exists (see
+// OptValueForKey).
+//
+//     myMap := map[string]map[string]int{"a": map[string]int{"b": 2}}
+//     matchers.In(t).Assert(myObject,
+//         matchers.ValueForKey("a").Should(
+//             matchers.ValueForKey("b").Should(Equal(2))))
```

Comment:
Is `myObject` supposed to be `myMap`?
Additionally I'm a bit confused by the syntax,
```golang
matchers.In(t).Assert(mMap, 
    matchers.ValueForKey("a").Should(
        matchers.ValueForKey("b").Should(Equal(2))
    )
)
````
Does this read as indicating that the value for `a` should be equal to another map, where the value for `b` should be equal to 2? Should there not be another call to `Equal` in there? Might be just misreading/misunderstanding it.

Reply:
It's not "the value for `a` should _be equal to another_ map where...", but rather, "the value for `a` should _be_ a map where...". The difference is that it's not making any assertions about any _other_ properties other than `b`.

In other words, it's not saying that `myMap` has to be `{ "a": { "b": 2 } }`. It could be `{ "a": { "x": true, "y": 7, "b": 2 } }`.

(but yes, `myObject` was supposed to be `myMap`)

Reply:
The value of this kind of thing might be clearer if I used an example other than `Equal` for the last expression. For instance, you could say `matchers.ValueForKey("a").Should(matchers.ValueForKey("b").Should(StringHasPrefix("x")))`, which would be true for `{"a": {"b": "x"}}` or `{"a": {"b": "xyz"}}`, but not for `{"a": {"b": "zyx"}}` or `{"a": {"c": "x"}}` or `{"a": true}`.

Reply:
The basic idea of these kinds of composable matchers is pretty close to the popular Hamcrest library in Java. There are some more examples of why it can be useful here: http://hamcrest.org/JavaHamcrest/tutorial